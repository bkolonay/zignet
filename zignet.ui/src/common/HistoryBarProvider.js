import moment from 'moment';

class HistoryBarProvider {

  /* 
    helped with percentage calculations:
     - https://www.calculatorsoup.com/calculators/math/percentage.php
     - https://www.timeanddate.com/date/timeduration.html
  */
  
	getErrorDivAttributes(historyBarWidth, now, testFailureDuration) {
		let oneDayAgo = moment(now.format()).subtract(24, 'hours');
    let testFailureStartTime = moment.utc(testFailureDuration.FailureStart).local();
    let minutesBetweenOneDayAgoAndTestFailureStartTime = moment.duration(testFailureStartTime.diff(oneDayAgo)).asMinutes();

    var failureDivStart = 0;
    if (minutesBetweenOneDayAgoAndTestFailureStartTime > 0) {
      let passingPercentOfLastDay = (minutesBetweenOneDayAgoAndTestFailureStartTime / 1440) * 100;
      failureDivStart = Math.floor((passingPercentOfLastDay / 100) * historyBarWidth);
    }
    else
      testFailureStartTime = oneDayAgo;

    var failureDivWidth = 0;
    if (testFailureDuration.FailureEnd) {
      let testFailureEndTime = moment.utc(testFailureDuration.FailureEnd).local();
      let failureDurationInMinutes = moment.duration(testFailureEndTime.diff(testFailureStartTime)).asMinutes();
      let failingPercentOfLastDay = (failureDurationInMinutes / 1440) * 100;
      failureDivWidth = Math.floor((failingPercentOfLastDay / 100) * historyBarWidth);
      if (failureDivWidth < 1)
      	failureDivWidth = 1;
    }
    else 
      failureDivWidth = historyBarWidth - failureDivStart;

    return {left: failureDivStart, width: failureDivWidth};
	}

  getDivTitle(testFailureDuration) {
    let testFailureStartTime = moment.utc(testFailureDuration.FailureStart).local();
    if (testFailureDuration.FailureEnd)
    {
      let testFailureEndTime = moment.utc(testFailureDuration.FailureEnd).local();
      let timeFromStartToEnd = testFailureStartTime.to(testFailureEndTime, true);
      return timeFromStartToEnd + ' from ' + testFailureStartTime.format('l h:mma') + ' - ' + testFailureEndTime.format('l h:mma');
    }
    else
    {
      let timeFromStartToEnd = testFailureStartTime.toNow(true);
      return timeFromStartToEnd + ' from ' + testFailureStartTime.format('l h:mma') + ' - now';
    }      
  }
}

export default HistoryBarProvider;