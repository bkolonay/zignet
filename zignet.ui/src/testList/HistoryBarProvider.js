import React from 'react';
import moment from 'moment';

function getFailureDivs(testFailureDurations, historyBarDiv) {
  const historyBarWidth = historyBarDiv.offsetWidth;
  let now = moment();
  var failureDivs = [];
  for (var i = 0; i < testFailureDurations.length; i++) {
    let failureDivAttributes = getErrorDivAttributes(historyBarWidth, now, testFailureDurations[i]);
    failureDivs.push(<div key={i} style={failureDivAttributes} title={getDivTitle(testFailureDurations[i])}/>);
  }
  return failureDivs;
}

  /* 
    helped with percentage calculations:
     - https://www.calculatorsoup.com/calculators/math/percentage.php
     - https://www.timeanddate.com/date/timeduration.html
  */
  
function getErrorDivAttributes(historyBarWidth, now, testFailureDuration) {
	let oneDayAgo = moment(now.format()).subtract(24, 'hours');
  let testFailureStartTime = moment.utc(testFailureDuration.FailureStart).local();
  let minutesBetweenOneDayAgoAndTestFailureStartTime = moment.duration(testFailureStartTime.diff(oneDayAgo)).asMinutes();

  var failureDivStart = 0;
  if (minutesBetweenOneDayAgoAndTestFailureStartTime > 0) {
    let passingPercentOfLastDay = (minutesBetweenOneDayAgoAndTestFailureStartTime / 1440) * 100;
    failureDivStart = Math.floor((passingPercentOfLastDay / 100) * historyBarWidth);
  }
  else
    testFailureStartTime = oneDayAgo;

  var failureDivWidth = 0;
  if (testFailureDuration.FailureEnd) {
    let testFailureEndTime = moment.utc(testFailureDuration.FailureEnd).local();
    let failureDurationInMinutes = moment.duration(testFailureEndTime.diff(testFailureStartTime)).asMinutes();
    let failingPercentOfLastDay = (failureDurationInMinutes / 1440) * 100;
    failureDivWidth = Math.floor((failingPercentOfLastDay / 100) * historyBarWidth);
    if (failureDivWidth < 1)
    	failureDivWidth = 1;
  }
  else 
    failureDivWidth = historyBarWidth - failureDivStart;

  return {left: failureDivStart, width: failureDivWidth};
}

function getDivTitle(testFailureDuration) {
  let testFailureStartTime = moment.utc(testFailureDuration.FailureStart).local();
  if (testFailureDuration.FailureEnd)
  {
    let testFailureEndTime = moment.utc(testFailureDuration.FailureEnd).local();
    let timeFromStartToEnd = testFailureStartTime.to(testFailureEndTime, true);
    return timeFromStartToEnd + ' from ' + testFailureStartTime.format('l h:mma') + ' - ' + testFailureEndTime.format('l h:mma');
  }
  else
  {
    let timeFromStartToEnd = testFailureStartTime.toNow(true);
    return timeFromStartToEnd + ' from ' + testFailureStartTime.format('l h:mma') + ' - now';
  }      
}

export { getFailureDivs, getErrorDivAttributes, getDivTitle };