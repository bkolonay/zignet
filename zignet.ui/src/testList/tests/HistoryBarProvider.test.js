import renderer from 'react-test-renderer';
import moment from 'moment';
import * as HistoryBarProvider from '../HistoryBarProvider'

// getFailureDivs
it('returns zero divs when no failures', () => {
  let historyBarDiv = {offsetWidth: 309}
  const testFailureDurations = [];

  let failureDivs = HistoryBarProvider.getFailureDivs(testFailureDurations, historyBarDiv);
  const tree = renderer.create(failureDivs).toJSON();
  expect(tree).toMatchSnapshot();
});

it('returns 1 div when 1 failure', () => {
  let historyBarDiv = { offsetWidth: 309 }
  let now = moment("2019-04-07T01:00:00-07:00");
  const testFailureDurations = [{
    FailureStart: moment("2019-04-06T03:00:00-07:00").utc().format(),
    FailureEnd: moment("2019-04-06T04:00:00-07:00").utc().format()
  }];

  let failureDivs = HistoryBarProvider.getFailureDivs(testFailureDurations, historyBarDiv, now);
  const tree = renderer.create(failureDivs).toJSON();
  expect(tree).toMatchSnapshot();
});

it('returns 2 divs when 2 failures', () => {
  let historyBarDiv = { offsetWidth: 309 }
  let now = moment("2019-04-07T01:00:00-07:00");
  const testFailureDurations = [{
    FailureStart: moment("2019-04-06T03:00:00-07:00").utc().format(),
    FailureEnd: moment("2019-04-06T04:00:00-07:00").utc().format()
  },
  {
    FailureStart: moment("2019-04-06T04:30:00-07:00").utc().format(),
    FailureEnd: moment("2019-04-06T05:00:00-07:00").utc().format()    
  }];

  let failureDivs = HistoryBarProvider.getFailureDivs(testFailureDurations, historyBarDiv, now);
  const tree = renderer.create(failureDivs).toJSON();
  expect(tree).toMatchSnapshot();
});

// getErrorDivAttributes()
it('gets a failure duration with a start and end time', () => {
  let now = moment("2019-04-07T01:00:00-07:00"); // 5/7/2019 1am
  var failureStart = moment("2019-04-06T03:00:00-07:00"); // 5/6/2019 3am
  var failureEnd = moment("2019-04-06T04:00:00-07:00"); // 5/6/2019 4am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(25);
  expect(errorDivAttributes.width).toEqual(12);
});

it('gets a failure with a duration of 1 second', () => {
  let now = moment("2019-04-07T01:00:00-07:00"); // 5/7/2019 1am
  var failureStart = moment("2019-04-06T03:00:00-07:00"); // 5/6/2019 3am
  var failureEnd = moment("2019-04-06T03:00:01-07:00"); // 5/6/2019 3:00:01am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(25);
  expect(errorDivAttributes.width).toEqual(1);
});

it('gets a failure duration with no end time', () => {
  let now = moment("2019-04-07T06:00:00-07:00"); // 5/7/2019 6am
  var failureStart = moment("2019-04-07T04:00:00-07:00"); // 5/7/2019 4am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(283);
  expect(errorDivAttributes.width).toEqual(26);
  expect(errorDivAttributes.left + errorDivAttributes.width).toEqual(309);
});

it('gets a failure duration with end time 1 second before now', () => {
  let now = moment("2019-04-07T03:00:00-07:00"); // 5/7/2019 3am
  var failureStart = moment("2019-04-07T02:00:00-07:00"); // 5/7/2019 2am
  var failureEnd = moment("2019-04-07T02:59:59-07:00"); // 5/7/2019 2:59:59am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(296);
  expect(errorDivAttributes.width).toEqual(12);
  expect(errorDivAttributes.width + errorDivAttributes.left).toEqual(308);
});

it('gets a failure duration with a start time before 24 hours ago and end time within 24 hours', () => {
  let now = moment("2019-04-07T01:00:00-07:00"); // 5/7/2019 1am
  var failureStart = moment("2019-04-06T00:30:00-07:00"); // 5/6/2019 12:30am
  var failureEnd = moment("2019-04-06T02:00:00-07:00"); // 5/6/2019 2am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(0);
  expect(errorDivAttributes.width).toEqual(12);
});

it('gets a failure duration with a start time before 24 hours ago and no end time', () => {
  let now = moment("2019-04-07T01:00:00-07:00"); // 5/7/2019 1am
  var failureStart = moment("2019-04-05T01:00:00-07:00"); // 5/5/2019 1am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(0);
  expect(errorDivAttributes.width).toEqual(309);
});

it('gets a failure duration with a start time 1 second before 24 hours ago and end time 1 second within 24 hours', () => {
  let now = moment("2019-04-07T01:00:00-07:00"); // 5/7/2019 1am
  var failureStart = moment("2019-04-06T00:59:59-07:00"); // 5/6/2019 12:59:59am (1 second before 24 hours ago)
  var failureEnd = moment("2019-04-06T01:00:01-07:00"); // 5/6/2019 1:00:01
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivAttributes = HistoryBarProvider.getErrorDivAttributes(309, now, testFailureDuration);

  expect(errorDivAttributes.left).toEqual(0);
  expect(errorDivAttributes.width).toEqual(1);
});

// getDivTitle()
it('get title when failure has a start and end time', () => {
  var failureStart = moment("2019-04-06T03:00:00-07:00"); // 5/6/2019 3am
  var failureEnd = moment("2019-04-06T04:00:00-07:00"); // 5/6/2019 4am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivTitle = HistoryBarProvider.getDivTitle(testFailureDuration);

  expect(errorDivTitle).toEqual('an hour from 5/6/2019 3:00am - 5/6/2019 4:00am');
});

it('get title when dates have double digits', () => {
  var failureStart = moment("2019-11-12T12:29:34-07:00"); // 11/12/2019 12:29pm
  var failureEnd = moment("2019-11-24T14:40:45-07:00"); // 11/24/2019 2:40pm
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: failureEnd.utc().format()
  };

  let errorDivTitle = HistoryBarProvider.getDivTitle(testFailureDuration);

  expect(errorDivTitle).toEqual('12 days from 11/12/2019 11:29am - 11/24/2019 1:40pm');
});

it('get title when end date is null', () => {
  var failureStart = moment("2019-04-06T03:00:00-07:00"); // 5/6/2019 3am
  let testFailureDuration = {
    FailureStart: failureStart.utc().format(),
    FailureEnd: null
  };

  let errorDivTitle = HistoryBarProvider.getDivTitle(testFailureDuration);

  expect(errorDivTitle).toMatch(new RegExp('5/6/2019 3:00am - now'));
});

it('does not throw when dates empty', () => {
  let testFailureDuration = {
    FailureStart: null,
    FailureEnd: null
  };

  let errorDivTitle = HistoryBarProvider.getDivTitle(testFailureDuration);

  expect(errorDivTitle).toBeDefined();
  expect(errorDivTitle).toBeTruthy();
});