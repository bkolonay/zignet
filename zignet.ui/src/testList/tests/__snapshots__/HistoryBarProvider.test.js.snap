// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns 1 div when 1 failure 1`] = `
<div
  style={
    Object {
      "left": 25,
      "width": 12,
    }
  }
  title="an hour from 5/6/2019 3:00am - 5/6/2019 4:00am"
/>
`;

exports[`returns 2 divs when 2 failures 1`] = `
Array [
  <div
    style={
        Object {
            "left": 25,
            "width": 12,
          }
    }
    title="an hour from 5/6/2019 3:00am - 5/6/2019 4:00am"
/>,
  <div
    style={
        Object {
            "left": 45,
            "width": 6,
          }
    }
    title="30 minutes from 5/6/2019 4:30am - 5/6/2019 5:00am"
/>,
]
`;

exports[`returns zero divs when no failures 1`] = `null`;
